### ISSUE TITLE
Implement PoissonVI to model scATAC-seq fragment counts

### ISSUE THREAD

**Issue by lauradmartens** [scverse/scvi-tools#1926]
**Is your feature request related to a problem? Please describe.**
PeakVI currently only supports modelling of binarized scATAC-seq data. In our [preprint](https://www.biorxiv.org/content/10.1101/2022.05.04.490536v1), we showed that it can be beneficial to actually model the scATAC-seq fragment counts.

**Describe the solution you'd like**
As previously discussed with @adamgayoso, it would be good to add the PoissonVI model to scvi, that users can also model the fragment count data.

**Comment by adamgayoso** [comment]
Hi @lauradmartens, can you please describe again the difference between scVI with a poisson likelihood and the poissonVI model?

**Comment by lauradmartens** [comment]
Hi Adam!

The model itself is similar but with some changes:
* AnnData setup checks that users input fragment counts and not binarized counts or read counts
* Function `reads_to_fragments` to approximately convert read counts to fragment counts
* Default architecture (number of layers etc.) and training procedure (learning rate etc.) is the one of PeakVI for ATAC data
* Function names as in PeakVI and adapted to ATAC terminology `get_accessibility_estimates`, `differential_accessibility`
* Possibility to output an accessibility probability from the Poisson rate parameter estimations in `get_accessibility_estimates`
*  And lastly but I have not implemented this yet, it might be interesting to exclude the region-specific bias (bias in the last layer) from the normalized accessibility estimates

But I am very open to also changing the structure or merging it with existing models if you think there is a more elegant solution.

**Comment by marianogabitto** [comment]
Laura,
      I have already implemented but never push the commit. Let me know what you want to do about this.

Mariano

**Comment by lauradmartens** [comment]
Hi Mariano, 

Cool to hear that! Sure, if this makes things easier! I am happy to help with input and documentation.

**Comment by adamgayoso** [comment]
I need to think more about it, but perhaps we can add the poisson functionality directly into PeakVI, and then have the option to create a PoissonPeakVI model that just inherits peakvi and sets poisson to true.

The second part would increase expose for people to cite your publication, but isn't strictly necesary code wise

**Comment by lauradmartens** [comment]
Thanks Adam! In this case, I think I would prefer the second option.

**Comment by marianogabitto** [comment]
In my code, I included in MultiVI. To be sincere, I did not see a major difference compared to PeakVI. A couple of thoughts (caveat, I read the paper and implemented a while ago so I might be wrong in my comments):
- Laura, can you paste the code to preprocess the reads before analysis?, I remember something about multiplying by 2 or dividing by 2 the number of reads per cell per peak.
- The poisson likelihood is cool because it has a normalized mean expression, independent from library size (like scvi). However, I remember that your implementation does not allow you to separate the regions factor from the mean expression. can you comment about this? (Basically, I am saying that to be able to get the normalized mean expression you need to correct for the region-specific factor)

**Comment by lauradmartens** [comment]
Yes, correct. As Adam said, the PoissonPeakVI would be for citation reasons. The main difference are in the decoder, the conversion step that you mention and the normalized accessibility. 

> * Laura, can you paste the code to preprocess the reads before analysis? I remember something about multiplying by 2 or dividing by 2 the number of reads per cell per peak.

In my pull request (#1927, I tagged you), I implemented the conversion code and also I check in setup_anndata that users don't provide the read counts or binarized data. Of course, this conversion is an approximation of the real number of fragments ber peak and ideally, users would count fragments in their preprocessing pipeline but for 10x read count matrices, this approximation works well and is very similar to actually counting fragments.


> * The poisson likelihood is cool because it has a normalized mean expression, independent from library size (like scvi). However, I remember that your implementation does not allow you to separate the regions factor from the mean expression. can you comment about this? (Basically, I am saying that to be able to get the normalized mean expression you need to correct for the region-specific factor)

That is nice, and completely correct. That was the point I was referring to here:

> * And lastly but I have not implemented this yet, it might be interesting to exclude the region-specific bias (bias in the last layer) from the normalized accessibility estimates

**Comment by adamgayoso** [comment]
I think to get things started it might be easier to work in pieces.

@lauradmartens would you be able to make a PR into `scvi.data` to add the reads to fragments as a preprocessing function?

https://github.com/scverse/scvi-tools/pull/1927#discussion_r1120435473

This can live in the `_preprocessing.py` file.

After that I just want to better understand how exactly your poisson model is written. It's my understanding that running scvi.model.SCVI with poisson likelihood is the same generative model as what you proposed, perhaps with a different architecture because peakvi uses different activation, number of units etc.

**Comment by marianogabitto** [comment]
Adam,
        there are simple things to do like modifying the _preprocessing.py file or making PeakVI chose between the Binary and the Poisson likelihood (I have done this actually). Should I start a new pull request with the data processing part alone? (tackling one issue at the time?

<

**Comment by adamgayoso** [comment]
I would first want to see if @lauradmartens would like to contribute this PR. Either way we should make it clear in that docstring that the procedure comes from the poisson vi paper

**Comment by lauradmartens** [comment]
Hi Adam, hi Mariano, sorry for not getting back earlier. I added the preprocessing function as you suggested @adamgayoso. As mentioned before I also wrote a function to check that the `anndata` object does actually contain read counts (more even than uneven counts) and non-binarized data. Should that also go in the preprocessing function?

**Comment by lauradmartens** [comment]
Hi, I wanted to revive this conversation because the PoissonVAE paper has now been accepted and it would be nice to refer to scvi for the model. @adamgayoso, could you tell me what is missing to include it in PeakVI? Would be happy to include it in the next PR.

**Comment by martinkim0** [comment]
Hi @lauradmartens, Adam graduated from the lab recently and is thus no longer actively developing scvi-tools. I'd be happy to help out with getting the model into the codebase. For now, I think it makes sense to implement a new model class under `scvi.external` that inherits from `PeakVI` and adds the necessary changes. I can provide more feedback once there's a PR.
